adr/keycloak-adr.md

```diff
@@ -0,0 +1,236 @@
+    H --> J
+    I --> J
+```
+The provided sequence diagram illustrates a scalable and decoupled architecture for handling user deprovisioning, leveraging NATS for centralized publish/subscribe messaging. This pattern ensures that when a user is deleted or deactivated in Keycloak, the event is published to NATS, from where it is distributed to various service listeners, including a GitLab Service Listener and an Audit Listener. The GitLab Service Listener is responsible for revoking the user's access in GitLab, while the Audit Listener updates the Admin UI to display the user deletion event. This architecture can be easily extended to include additional services by adding more listeners that subscribe to the relevant NATS topics for user status changes.
+```mermaid
+sequenceDiagram
+    participant Keycloak as "Keycloak"
+    participant EventListener as "Keycloak Event Listener"
+    participant NATS as "NATS Messaging System"
+    participant GitLabService as "GitLab Service Listener"
+    participant UIService as "Audit Listener"
+    participant AdminUI as "Admin UI"
+    participant OtherService as "Other Service Listener"
+
+    Keycloak->>EventListener: User Deletion/Deactivation
+    EventListener->>NATS: Publish User Status Change
+    NATS->>GitLabService: Notify User Deletion
+    NATS->>UIService: Update for UI Display
+    NATS->>OtherService: Notify User Deletion (Extendable)
+    GitLabService->>GitLab: Revoke User Access
+    UIService->>AdminUI: Display User Deletion Event
+    OtherService->>OtherSystem: Revoke User Access (Extendable)
+    Note over Keycloak,AdminUI: De-provisioning with Centralized Pub/Sub Messaging
+    Note over OtherService,OtherSystem: Similar pattern can be extended to other services
+```
+This diagram highlights the flexibility and scalability of using NATS for event-driven communication across different services within an organization's infrastructure. By adding more service listeners (e.g., "Other Service Listener") for additional systems (e.g., "OtherSystem"), this architecture allows for seamless integration and handling of user deprovisioning events across a wide array of services, ensuring a consistent and secure approach to access management and compliance.
+
+#### Security between Services
+The security between services is a critical aspect of the proposed architecture. The communication between Keycloak, the event listener, NATS, and the various service listeners must be secure to ensure the integrity and confidentiality of user deprovisioning events. The following security measures will be implemented:
+- **TLS/SSL**: All communication between Keycloak, the event listener, NATS, and the service listeners will be encrypted using TLS/SSL to ensure data confidentiality and integrity. This will be achieved by utilizing istio for mutual TLS between services. This will  Strict mTLS will be enforced between services to ensure that only authorized entities can communicate with each other.
```

Comment:
```suggestion
- **TLS/SSL**: All communication between Keycloak, the event listener, NATS, and the service listeners will be encrypted using TLS/SSL to ensure data confidentiality and integrity. This will be achieved by utilizing istio for mutual TLS between services. Strict mTLS will be enforced between services to ensure that only authorized entities can communicate with each other.
```